### PR TITLE
feat: Implements Popular Dashboards, when index dashboard flag is enabled

### PR DESCRIPTION
**Pull request by verdan**
### Summary of Changes

This PR implements the popular dashboards when the index dashboard flag is enabled in the frontend configs. This is backward compatible with the metadata endpoint. We will keep the metadata endpoint for popular tables with a deprecation warning.
The UI leverages all of the existing components and does not introduce any new components. 

Screenshot 2021-06-24 at 15 03 35

Screenshot 2021-06-24 at 14 59 31

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: frontend/amundsen_application/static/js/components/PopularResources/index.tsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+export interface DispatchFromProps {
+  getPopularResources: () => GetPopularResourcesRequest;
+}
+
+export type PopularResourcesProps = StateFromProps & DispatchFromProps;
+
+export class PopularResources extends React.Component<PopularResourcesProps> {
+  componentDidMount() {
+    this.props.getPopularResources();
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/PopularResources/index.tsx:47]:
```suggestion
    const {getPopularResources} = this.props;
    
    getPopularResources();
```

File: frontend/amundsen_application/static/js/components/PopularResources/index.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      <PaginatedResourceList
+        allItems={popularResources}
+        emptyText={EMPTY_POPULAR_RESOURCES_MESSAGE}
+        itemsPerPage={POPULAR_RESOURCES_PER_PAGE}
+        source={POPULAR_RESOURCES_SOURCE_NAME}
+      />
+    );
+  };
+
+  generateTabKey = (resource: ResourceType) => `popularresourcetab:${resource}`;
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/PopularResources/index.tsx:67]: I would move this out of the class

File: frontend/amundsen_application/static/js/components/PopularResources/index.tsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  generateTabTitle = (resource: ResourceType): string => {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/PopularResources/index.tsx:69]: I would pass here the popularResources prop and move this function out of the class

File: frontend/amundsen_application/static/js/components/PopularResources/index.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  isLoaded: state.popularResources.popularResourcesIsLoaded,
+});
+
+export const mapDispatchToProps = (dispatch: any) =>
+  bindActionCreators({ getPopularResources }, dispatch);
+
+export default connect<StateFromProps, DispatchFromProps>(
+  mapStateToProps,
+  mapDispatchToProps
+)(PopularResources);
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/PopularResources/index.tsx:138]: I assume we are reworking the old version right? If not, I would push this to be a functional component.

**Review by Golodhros — COMMENTED** [review]
Looks good in general, only some minor comments.

Is this 100% backwards compatible?

**Comment by verdan** [comment]
> Looks good in general, only some minor comments.
> 
> 
> 
> Is this 100% backwards compatible?

I need to refactor the python config flags, and trying to figure out how to make it work with existing configurations. Apart from that the endpoints in metadata are backwards compatible.

File: frontend/amundsen_application/static/js/components/PopularResources/index.tsx (hunk 4 above)
Comment (verdan) [reply frontend/amundsen_application/static/js/components/PopularResources/index.tsx:142]: we are indeed re-using the existing component.

**Comment by feng-tao** [comment]
what is the heuristic to define popular dashboard?

**Comment by verdan** [comment]
Leverages the same formula (and even function) we have for the popular tables. i.e., logarithm on the total number of reads.

File: metadata/metadata_service/entity/resource_type.py [hunk 5]
```diff
@@ -5,6 +5,9 @@
 
 
 class ResourceType(Enum):
+    """
+    DEPRECATED, Please use amundsen_common/entity/resource_type.py instead
```
Comment (dorianj) [inline metadata/metadata_service/entity/resource_type.py:9]: Could we import it and do `ResourceType = amundsen_common.entity.ResourceType` instead?

**Review by dorianj — APPROVED** [review]

**Comment by dorianj** [comment]
Great generalization!

**Comment by feng-tao** [comment]
> Leverages the same formula (and even function) we have for the popular tables. i.e., logarithm on the total number of reads.

I didn't look the code but wonder whether we could the heuristic pluggable. I am pretty sure some BI dashboard tool doesn't have read count information (e.g redash) which won't make popular dashboard useful.

**Comment by verdan** [comment]
> > Leverages the same formula (and even function) we have for the popular tables. i.e., logarithm on the total number of reads.
> 
> I didn't look at the code but wonder whether we could make the heuristic pluggable. I am pretty sure few BI dashboard tool doesn't have read count information (e.g redash) which won't make popular dashboard useful.

This is a good point. I have added an issue for this to keep this in sight and also to keep the scope of this PR limited.

**Pull request merged**